src/engine/Match_search_mode.ml

```diff
@@ -554,147 +554,209 @@ let children_explanations_of_xpat (env : env) (xpat : Xpattern.t) : ME.t list =
 
 let hook_pro_entropy_analysis : (string -> bool) option ref = ref None
 
-let rec filter_ranges (env : env) (xs : (RM.t * MV.bindings list) list)
-    (cond : R.metavar_cond) : (RM.t * MV.bindings list) list =
+(* [filter_ranges] filters all our ranges by all our conditions simultaneously.
+   We do this via "phases" -- first, for each condition, and within each
+   condition, evolving each active range simultaneously.
+*)
+let rec filter_ranges (env : env) (init_ranges : RM.t list)
+    (conds : (Tok.t * R.metavar_cond) list) : RM.ranges * ME.t option list =
+  let ranges_and_bindings, expls =
+    List.fold_left
+      (fun (acc_ranges, acc_expls) (tok, cond) ->
+        let new_ranges =
+          List_.map_filter
+            (fun (range, bindings) ->
+              match filter_single env range cond with
+              | None -> None
+              | Some new_bindings -> Some (range, new_bindings :: bindings))
+            acc_ranges
+        in
+        let new_expl =
+          if_explanations env (List_.map fst new_ranges) []
+            (OutJ.Filter (Tok.content_of_tok tok), tok)
+        in
+        (new_ranges, new_expl :: acc_expls))
+      (List_.map (fun r -> (r, [])) init_ranges, [])
+      conds
+  in
+  (* Each entry in bindings_per_cond is a list, each entry of which corresponds to the
+     bindings which may have been produced from that condition.
+     There are many such bindings because we may have multiple possibilities. For
+     instance, a metavariable-regex may potentially match numerous times, each one
+     introducing a different set of bindings.
+
+     For metavariable-regex foo\((?<A>)(?<B>)\) on the text "foo(12)", we could have
+     [ [A, ""; B, "12"]
```

Comment:
I would have appreciated using a running example, a bit like in here, to better visualize what each step is doing.